### DISCUSSION TITLE
M73 - Set Print Progress

### DISCUSSION THREAD

**Discussion by synman** [synman/Ender-3-S1-Pro-Firmware#31] [Features]
Touchscreen % complete and time remaining should be updatable on the touchscreen when external senders support it

**Comment by synman** [comment]
M73 in action using the <a href="https://plugins.octoprint.org/plugins/m73progress/">M73 Progress</a> plug-in for Octoprint:

![PXL_20230212_052537166](https://user-images.githubusercontent.com/1299716/218295326-3d5503e0-6781-493c-aaee-fccd5277a9d6.jpg)

**Comment by synman** [comment]
I delivered this functionality to the bugfix-2.1.x branch this morning and released the binaries for it there.

I'll cycle back to 2.1.2 and 2.0.8 later today or tomorrow.

**Reply by JohnScottUK** [reply]
That would be great. 2.1.2 is working well for me. Updated from V2.0.8.26 and Screen V1.0.3 and it has been OK with some bed levelling and benchy. Tried adding in M73 and tried M75/M77 start/stop timers, but %age etc does not update.

Looking forward to progress bars on LCD :-)

**Reply by synman** [reply]
I may get to this tonight.   I was down most of the week as my MBP's logic board fried and I had to make an Apple Store run to drop it off repairs.   

Has taken the better part of the last 3 days to get everything re-installed / sorted out.

**Reply by lipa46** [reply]
the progress is not working with screen version 1.0.3 and latest firmware bug_fix_2.1.x UBL100.

**Reply by synman** [reply]
@JohnScottUK M73 is available now in 2.1.2 (untested).

@lipa46 please confirm you have the Octoprint M73 Progress plugin installed and active.  M73 does not work on its own.  Something has to actually send the updates.

**Reply by synman** [reply]
> the progress is not working with screen version 1.0.3 and latest firmware bug_fix_2.1.x UBL100.

As suspected, working just fine.  Perhaps you missed the plugin requirement?

![PXL_20230226_190255529.jpg](https://user-images.githubusercontent.com/1299716/221432807-684302d0-325f-4e65-be34-59506db8e5c9.jpg)

**Reply by synman** [reply]
Complimentary Octoprint status...

![Screenshot_20230226-140235.png](https://user-images.githubusercontent.com/1299716/221434679-9a7e118f-d9e1-4b92-bb9c-f75d4f9f7c24.png)

**Reply by lipa46** [reply]
Sorry my bad. I did not see that it need a plugin. I will try it tomorrow.

**Reply by synman** [reply]
> Sorry my bad. I did not see that it need a plugin. I will try it tomorrow.

Please start reading the details in the threads here before opening issues / creating discussions or I'm gonna have to start charging you for support time via my sponsors link.

**Comment by JohnScottUK** [comment]
I have an odd problem with M73. When printing via USB, the %age keeps jumping to 100% then back to the expected %age.

I am printing via USB (using a Mintion BeaglePrint camera). I have used a Cura post-processing script that added M73 P0 through M73 P100 and M73 Rn down to M73 r0 statements through the gcode. There is no option for BeaglePrint camera to inject them like OctoPrint M73 progress plugin does. They are added separately by Cura script. I have checked the gcode and there is only one M73 P100 statement at the end of the file. I can see 101 M73 Px statements in the file (from 0 to 100) and 68 M73 Rx statements in the file from M73 R17 to M73 R0 (each statement is output every 15 seconds, so there are typically 4 M73 Rx statements for each value of x (M73 R15 appears 4 times for example).

Not sure why progress would jumpt to P100 and then back to X % as per M73. I do note that the M73 command at https://marlinfw.org/docs/gcode/M073.html shows P as mandatory and R as optional. Not sure if this has an impact when M73 Rx is generated without a Px value.

Also whilst the hourglass shows decreasing time, the clock does not show progress. How is that driven? Should I be using M75 or similar to start a job timer in my gcode?

**Reply by JohnScottUK** [reply]
OK. I extracted all the M73 statements from my flow cube. I think when an M73 command is sent without a P argument, the %age updates with a random value (uninitialised memory for the P value)?
[Progress.gcode.txt](https://github.com/synman/Ender-3-S1-Pro-Firmware/files/10855720/Progress.gcode.txt)

**Reply by synman** [reply]
I wouldn't be surprised if it's an initialization issue.  The screen handler does direct memory copies of it's buffer touchscreen and I've seen oddities when you don't null terminate strings etc in preparation for populating it.

I'll see what I can see given what you provided.

Thanks!

**Reply by JohnScottUK** [reply]
I think its missing braces around here https://github.com/synman/Marlin/blob/706174edf5feb837b7f2a7f29d03be06e95219f7/Marlin/src/gcode/lcd/M73.cpp#L63

If parser.seenval('P') is false then only line 65/66 are not executed. It will always go on to ui.set_progress on line 69.

I think it should look like this (formatted a bit for clarity):
```cpp
    #if ENABLED(SET_PROGRESS_PERCENT)
      if (parser.seenval('P')) { //<-- missing brace.
        #if ENABLED(CREALITY_TOUCHSCREEN)
          if (parser.value_byte() == 0)
            last_start_time = HAL_GetTick();
        #endif

        ui.set_progress((PROGRESS_SCALE) > 1
          ? parser.value_float() * (PROGRESS_SCALE)
          : parser.value_byte()
        );
      } //<-- Missing brace.
    #endif
```

**Reply by synman** [reply]
`ENABLED(CREALITY_TOUCHSCREEN)` is a precompiler macro as is `ENABLED(SET_PROGRESS_PERCENT)`.  assuming both of these variables are defined (which they are) the code that gets delivered to the compiler looks like this:
```
      if (parser.seenval('P'))
          if (parser.value_byte() == 0)
            last_start_time = HAL_GetTick();

        ui.set_progress((PROGRESS_SCALE) > 1
          ? parser.value_float() * (PROGRESS_SCALE)
          : parser.value_byte()
        );
```

See how the two `ifs` latch together?   I typically don't do shorthand for this reason, but in this case it is working as it should.

**Reply by synman** [reply]
I've created an issue for this for us to track it rather than fill up the Features thread like what myself and a few others did for UBL lol :)

https://github.com/synman/Ender-3-S1-Pro-Firmware/issues/39

**Reply by synman** [reply]
And I take back what I said about the latching ifs.  my eyes failed to look at the next line.  you are correct.

**Reply by synman** [reply]
this should be fixed now

**Reply by JohnScottUK** [reply]
Will test it. I always use braces even round single-line if statements. Number of times I've started at code unable to work out why its not working.

**Reply by synman** [reply]
I used your gcode as my test criteria so at least we know it'll work
